### PR TITLE
Refactor: BaseInvenTreeSetting

### PR DESCRIPTION
**Pull request by wolflu05**
This PR tries to add typings for settings and make the `common.BaseInvenTreeSetting` model more generic. This simplifies the implementation for the #4824 , see https://github.com/inventree/InvenTree/pull/4824#discussion_r1193650747

Actually first, I planned to only add a `get_filters` method like the `get_kwargs` method until I now discovered that something like this was already there onetime and was removed. (See: 5ee9af7f0ebed1cb56ea488299824daf0db0afb8). Let me know if you have a better idea to make it more generic.

I tested this and discovered no bugs or issues that something doesnt work or break with this PR. Hopefully tests cover the other half 😄.

### CODE REVIEW

**Comment by matmair** [comment]
@wolflu05 are you sure that the subclass checking is not required for plugin settings?

**Review by matmair — COMMENTED** [review]
Looks good in general

File: InvenTree/plugin/models.py
Comment (matmair) [inline InvenTree/plugin/models.py:1]: Line 171 lose coverage with the change - https://coveralls.io/builds/60021910/source?filename=InvenTree%2Fplugin%2Fmodels.py#L171

**Review by matmair — CHANGES_REQUESTED** [review]

**Comment by wolflu05** [comment]
> are you sure that the subclass checking is not required for plugin settings?

I used project wide code search and checked all method calls individually where I changed something if there is a `plugin` kwarg. And then looked through the list for `plugin=` results, changed everything there what I found to call the `.plugin_config()` function there.

**Comment by matmair** [comment]
So we can remove that section entirely?

**Comment by wolflu05** [comment]
What section? The subclass checking in the BaseInventreeSettingClass? Yes. Ideally everything is moved out now and should be implemented in new code like this.

Do you want me to add general tests in this PR? But how do we test an abstract model?

File: InvenTree/plugin/models.py
Comment (wolflu05) [reply InvenTree/plugin/models.py:1]: Removed that line, as it is not needed anymore as it only gets a `PluginConfig` now.

**Comment by matmair** [comment]
Something seems off with the coverage - I will let it run again

**Review by matmair — COMMENTED** [review]
Looks good but not sure if the uniqueness will work this way

**Comment by wolflu05** [comment]
How do you mean that?

**Review by SchrodingersGat — APPROVED** [review]

**Comment by matmair** [comment]
> How do you mean that?

I was worried about removing the kwargs passed to the uniqueness check for user settings, in local testing it seems to work this way though.

**Review by matmair — APPROVED** [review]

**Comment by matmair** [comment]
@SchrodingersGat 0.12.0 is looking like a big change under the hood - I think especially this one needs a bit of testing in the real world before we release a stable

**Comment by SchrodingersGat** [comment]
@wolflu05 thanks for the contribution

@matmair agreed, we won't be pushing out this release immediately!

**Pull request merged**